File: CSE101/pa4/List.c
////
////            Programming Assignment #4
////  Breadth First Search and Shortest Paths in Graphs
////        Created by: Nicholas Scagliotti
////                nscaglio@ucsc.edu
////                   CMPS 101
////                   Fall '18
////
#include<stdio.h>
#include<stdlib.h>
#include "List.h"


typedef struct NodeObj{						//Struct Node Object
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{						//Struct List Object
	Node front, back, cursor;
	int length;
	int index;
} ListObj;

int isEmpty(List L){							//Checks to see if list is empty
	if (L == NULL){							//Returns true if it's empty, false otherwise
		printf("Error: Called isEmpty() function while list reference is NULL.");
		exit(1);
	}
	return L->length == 0;						//Empty list has length = 0
}

//
//Constructors/Destructors
//

Node newNode(int data){						//Node Constructor
	Node n = malloc(sizeof(NodeObj));
	n->data = data;
	n->next = NULL;
	n->prev = NULL;
	return (n);
}

void freeNode(Node* pN){						//Node Destructor
	if (pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}

List newList(void){							//List Constructor
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = L->cursor = NULL;
	L->length = 0;
	L->index = -1;
	return (L);
}

void freeList(List* pL){						//List Destructor
	if (pL != NULL && *pL != NULL){
		Node n = (*pL)->front;
		while(n != NULL){
			Node temp = n;
			n = n->next;
			free(temp);
		}
		free(*pL);
		*pL = NULL;
	}
}

//
//Access Functions
//

int length(List L){							//Returns the length of the list
	if (L == NULL){
		printf("Error: Called length() function while list reference is NULL. \n");
		exit(1);
	}
	return L->length;
}

int index(List L){								//Returns the current index
	if (L == NULL){
		printf("Error: Called index() function while list reference is NULL. \n");
		exit(1);
	}
	return L->index;
}

int front(List L){								//Returns the value of the element in the front of the list
	if (L == NULL){
		printf("Error: Called front() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called front() while list is empty. \n");
		exit(1);
	}
	return L->front->data;
}

int back(List L){								//Returns the value of the element in the back of the list
	if (L == NULL){
		printf("Error: Called back() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called back() while list is empty. \n");
		exit(1);
	}

	return L->back->data;
}

int get(List L){								//Returns the value of the element which the cursor is on
	if (L == NULL){
		printf("Error: Called get() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called get() while list is empty. \n");
		exit(1);
	}
	if(L->cursor == NULL){
		printf("Error: Called get() function while cursor is set to NULL. \n");
		exit(1);
	}
	return L->cursor->data;
}

int equals(List A, List B){					//Checks to see if both lists are the same sequence
	Node X = NULL;								//Returns true if they match, false otherwise
	Node Y = NULL;
	int flag = 1;

	if (A == NULL || B == NULL){
		printf("Error: Called equals() while list reference is NULL. \n");
		exit(1);
	}

	X = A->front;
	Y = B->front;

	if (A->length != B->length){
		return 0;
	}
	while (flag && X != NULL && Y != NULL){
		flag = (X->data == Y->data);
		X = X->next;
		Y = Y->next;
	}
	return flag;
}

//
//Manipulation Procedures
//

void clear(List L){							//Clears the List, making it an empty list
	//Node n = L->front;
	if (L == NULL){
		printf("Error: Called clear() function while list reference is NULL. \n");
		exit(1);
	}
	while (!(isEmpty(L))){
		deleteFront(L);
		/*Node curr = n;
		n = n->next;
		free(curr);*/
	}
	L->cursor = NULL;
	L->front = NULL;
	L->back = NULL;
	L->length = 0;
	L->index = -1;
}

void moveFront(List L){						//Places the cursor under the front element
	if (L == NULL){
		printf("Error: Called moveFront() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called moveFront() function while list is empty. \n");
		exit(1);
	}
		L->cursor = L->front;
		L->index = 0;
}

void moveBack(List L){							//Places the cursor under the back element
	if (L == NULL){
		printf("Error: Called moveBack() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called moveBack() function while list is empty. \n");
		exit(1);
	}
		L->cursor = L->back;
		L->index = L->length - 1;
}

void movePrev(List L){							//Moves the cursor towards the front of of the list
	if (L == NULL){
		printf("Error: Called moveBack() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called moveBack() function while list is empty. \n");
		exit(1);
	}
	if (L->index > 0 && L->index <= L->length - 1){
		L->cursor = L->cursor->prev;
		L->index--;
	}
	else if (L->index <= 0){					//If cursor is already on the last element, cursor becomes undefined
		L->cursor = NULL;
		L->index = -1;
	}
}

void moveNext(List L){							//Moves the cursor towards the back of the list
	if (L == NULL){
		printf("Error: Called moveNext() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called moveNext() function while list is empty. \n");
		exit(1);
	}
	if (L->index != L->length - 1 && L->cursor != NULL){
		L->cursor = L->cursor->next;
		L->index++;
	}
	else if (L->index == L->length - 1 || L->cursor != NULL){	//If cursor is already on the last element, cursor becomes undefined
		L->cursor = NULL;
		L->index = -1;
	}
}

void prepend(List L, int data){				//Inserts new element ahead of the current front element
	Node n = newNode(data);

	if (L == NULL){
		printf("Error: Called prepend() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		L->front = L->back = n;
		L->cursor = L->front;
	}
	else{
		L->front->prev = n;
		n->next = L->front;
		L->front = n;
		L->index++;
	}
	L->length++;
}

void append(List L, int data){					//Inserts new element behind the current back element
	Node n = newNode(data);
	if (L == NULL){
		printf("Error: Called append() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		L->front = L->back = n;
	}
	else{
		L->back->next = n;
		n->prev = L->back;
		L->back = n;
	}
	L->length++;
}

void insertBefore(List L, int data){			//Inserts a new element before the current cursor
	Node n = newNode(data);
	if (L == NULL){
		printf("Error: Called insertBefore() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called insertBefore() function while list is empty. \n");
		exit(1);
	}
	if (L->cursor == NULL){
		printf("Error: Called insertBefore() function while cursor is NULL. \n");
		exit(1);
	}
	if (L->cursor == L->front){
		prepend(L, data);						//If cursor is on front element, calling the prepend function saves time
	}
	else{
		n->next = L->cursor;
		n->prev = L->cursor->prev;
		L->cursor->prev->next = n;
		L->cursor->prev = n;
		L->index++;
		L->length++;
	}
}


void insertAfter(List L, int data){			//Inserts a new element after the current cursor
	Node n = newNode(data);
	if (L == NULL){
		printf("Error: Called insertAfter() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called insertAfter() function while list is empty. \n");
		exit(1);
	}
	if (L->cursor == NULL){
		printf("Error: Called insertAfter() function while cursor is NULL. \n");
		exit(1);
	}
	if (L->cursor == L->back){					//If cursor is on back element, calling the append function saves time
		append(L, data);
	}
	else{
		n->next = L->cursor->next;
		n->prev = L->cursor;
		L->cursor->next->prev = n;
		L->cursor->next = n;
		L->length++;
	}
}

void deleteFront(List L){						//Deletes the front node of the list
	Node n = L->front;
	if (L == NULL){
		printf("Error: Called deleteFront() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called deleteFront() function while list is empty. \n");
		exit(1);
	}
	if (L->length == 1){				//Case for a list with only one element
		L->cursor = L->back = L->front = NULL;
		L->index = -1;
		L->length = 0;
		freeNode(&n);
	}
	else{
		L->front->next->prev = NULL;
		L->front = L->front->next;
		L->length--;
		freeNode(&n);
	}
}

void deleteBack(List L){						//Deletes the back node of the list
	Node n = NULL;
	if (L == NULL){
		printf("Error: Called deleteBack() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called deleteBack() function while list is empty. \n");
		exit(1);
	}
	n = L->back;
	if (L->length == 1){					//Case for a list with only one element
		L->front = L->back = L->cursor = NULL;
		L->index = -1;
		L->length = 0;
		freeNode(&n);
	}
	else{
		L->back->prev->next = NULL;
		L->back = L->back->prev;
		if(L->cursor != NULL){
			L->index--;
		}
		L->length--;
		freeNode(&n);
	}
}

void delete(List L){							//Deletes the node under the current cursor
	Node n = NULL;
	if (L == NULL){
		printf("Error: Called delete() function while list reference is NULL. \n");
		exit(1);
	}
	if (isEmpty(L)){
		printf("Error: Called delete() function while list is empty. \n");
		exit(1);
	}
	if (L->cursor == NULL){
		printf("Error: Called delete() function while cursor is NULL. \n");
		exit(1);
	}
	if (L->cursor == L->back){					//Case for deleting back node
		deleteBack(L);
	}
	else if (L->cursor == L->front){			//Case for deleting front node
		deleteFront(L);
	}
	else{
		n = L->cursor;
		L->cursor->prev->next = L->cursor->next;
		L->cursor->next->prev = L->cursor->prev;
		L->cursor = NULL;
		L->index = -1;
		L->length--;
		freeNode(&n);
	}
}

//
//Other Operations
//

void printList(FILE *out, List L){				//Function to print the list
	Node n = L->front;

	if (L == NULL){
		printf("Error: Called printList() function while list reference is NULL. \n");
		exit(1);
	}
	while (n != NULL){
		printf("%d", n->data);
		n = n->next;
	}
}

List copyList(List L){							//Copies current list to a new list
	List C = newList();
	Node n = L->front;
	while (n != NULL){
		append(C, n->data);
		n = n->next;
	}
	return C;
}

File: CSE101_Data-Structures-and-Algorithms/pa1/List.java
        ////
        ////
        ////            Programming Assignment #1
        ////                Integer List ADT
        ////        Created by: Nicholas Scagliotti
        ////                nscaglio@ucsc.edu
        ////                   CMPS 101
        ////                   Fall '18
        ////
        ////
public class List{

	private class Node{
		int data;
		Node next, prev;

		Node(int data){                                         //Node Constructor
			this.data = data;
			next = null;
			prev = null;
		}
		
	}

	private Node front, back, cursor;				//Fields to maintain desired information
	private int length, index;

	List() {										// List Constructor
            front = back = cursor = null; 
            length = 0; 
            index = -1;
        }	
        
	boolean isEmpty(){						//Checks to see if the list is empty. True if it's empty, false if it's not.
		return length == 0 && index < 0;                           //Len should be == 0 and index should be < 0.
	}

		//
		//			Access Functions
		//

	boolean equals(List L){                                         //Checks to see if List and L are the same sequence of integers.
		boolean flag = true;                                    //Returns true if they match, false if they don't.
		Node a  = this.front;
		Node b  = L.front;

		if(this.length == L.length){
			while(flag && a != null){
				flag = (a.data == b.data);
				a = a.next;
				b = b.next;
			}
			return flag;
		}
		else{
			return false;
		}
	}

      
	int length(){							//Returns the length of the array
		return length;
	}

	int index(){							//Returns the current index
		return index;
	}

	int front(){							//Returns the value of the element in the front of the array
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called front() while list is empty.");       //Error case for an empty list
		}
		return front.data;
	}

	int back(){							//Returns the value of the element in the back of the array
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called back() while list is empty.");
		}
		return back.data;
	}

	int get(){					//Returns the value of the element which the cursor is on
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called get() while list is empty.");
		}
		if(cursor == null){
			throw new RuntimeException("Error: Called get() while cursor is undefined.");   //Error case for a null cursor
		}
		return cursor.data;
	}

		//
		//			Manipulation Procedures
		//

	void clear(){							//Clears the List, making it an empty list
		front = back = cursor = null;
		length = 0;
		index = -1;
	}

	void moveFront(){						//Places the cursor under the front element
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called moveFront() while list is empty.");
		}
		else{
			cursor = front; //0
			index = 0;
		}
	}

	void moveBack(){						//Places the cursor under the back element
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called moveBack() while list is empty.");
		}
		else {
			cursor = back; // length - 1
			index = length - 1;
		}
	}
	void movePrev(){						//Moves the cursor towards the front of of the list.
		if(index == 0 || index == -1){                          //If cursor is already on the first element, cursor becomes undefined.
			cursor = null;
			index = -1;
		} 
		else if(index > 0 && index <= length -1){
			cursor = cursor.prev;
			index--;
		}
	}

	void moveNext(){						//Moves the cursor towards the back of the list.
		if(index == length - 1 || index == -1){                    //If cursor is already on the last element, cursor becomes undefined.
			cursor = null;
			index = -1;
		}
		else if(index >= 0 && index < length - 1){
			cursor = cursor.next;
			index++;
		}
	}
	

	void prepend(int data){						//Inserts new element ahead of the current front element 
		Node n = new Node(data);

		if(this.isEmpty()){
			front = back = n;
		}
		else{
			front.prev = n;
			n.next = front;
			front = n;
			index++;
		}
		length++;
	}

	void append(int data){                                          //Inserts new element behind the current back element
		Node n = new Node(data);
		if(this.isEmpty()){
			front = back = n;
		}
		else{
			back.next = n;
			n.prev = back;
			back = n;
		}
		length++;
	}

	void insertBefore(int data){                                    //Inserts a new element before the current cursor
		Node n = new Node(data);

		if(this.isEmpty()){
			throw new RuntimeException("Error: Called insertBefore() while list is empty.");
		}
		else if(cursor == null){
			throw new RuntimeException("Error: Called insertBefore() while cursor is undefined.");
		}
		else if(cursor == front){                               //If cursor is on front element, calling the prepend function saves time and prevents any issues
			prepend(data);
		}
		else{
			n.next = cursor;
			n.prev = cursor.prev;
			cursor.prev.next = n;
                        cursor.prev = n;
			index++;
			length++;
		}

	}

	void insertAfter(int data){                                    //Inserts a new element after the current cursor
		Node n = new Node(data);

		if(this.isEmpty()){
			throw new RuntimeException("Error: Called insterAfter() while list is empty.");
		}
		else if(cursor == null){
			throw new RuntimeException("Error: Called insertAfter() while cursor is undefined.");
		}
		else if(cursor == back){
			append(data);                                  //If cursor is on front element, calling the append function saves time and prevents any issues
		}
		else{
			
			n.next = cursor.next;
			n.prev = cursor;
                        cursor.next.prev = n;
			cursor.next = n;
			length++;
		}
	}

	void deleteFront(){
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called deleteFront() while list is empty.");
		}
		else if(this.length == 1){                     //Single case for a list with only one element
                        front = back = cursor = null;
			index = -1;
			length--;
		}
		else{                                       //Case for all lists with length > 1
                        front.next.prev = null;
			front = front.next;
			length--;
		}
	}

	void deleteBack(){
		if(this.isEmpty()){
                        throw new RuntimeException("Error: Called deleteBack() while list is empty.");
		}
                else if(this.length == 1){                     //Single case for a list with only one element
                        front = back = cursor = null;
			index = -1;
			length--;
                }
		else{                                       //Case for all lists with length > 1
			back.prev.next = null;
			back = back.prev;
			length--;
		}
	}

	void delete(){                                          
		if(this.isEmpty()){
			throw new RuntimeException("Error: Called delete() while list is empty or cursor is undefined.");
		}
		else{
			if(cursor.prev != null){
				cursor.prev.next = cursor.next;
			}
			if(cursor.next != null){
				cursor.next.prev = cursor.prev;
			}
			cursor = null;
			length--;
		}
	}
        
        //
        //          Other Methods
        //
        
        @ Override public String toString(){			//Overrides Object's toString() method, returning a string representation of the list
                if(this.isEmpty()){
                      return "Error: Called print function while list is empty.";      
                }
                else{
                    String string = "";
                    Node n = front;
                    for(int i = 0; i < length; i++){
                        string += n.data +" ";
                        n = n.next;
                    }
                    return string;
                }
	}
        
        List copy(){                                            //Returns a copy of the desired list
		List copy = new List();
		Node n = this.front;
		while(n != null){
			copy.append(n.data);
			n = n.next;
		}
                return copy;
	}
}